### PR TITLE
Implement Callback Providers for dynamic file lists

### PR DESCRIPTION
**Pull request by robertscherer**
As discussed with @markstory on IRC today, it would be nice to be able to add callback strings to the INI configuration, which modify the files[] list dynamically.

This is a first, naive implementation of how I would imagine this could work.

I'm mainly unsure about the position in the process the conversion from a callback string to a file list should happen.

I have implemented this in AssetConfig which is the soonest point to do this - which I think is a sensitive choice, but I could be wrong. 

An alternative would be moving this into the asset/file scanning part of the process, but this could have unintended side effects.

Also, I'm unsure about how to handle errors - should Exceptions be thrown if the callback isn't callable or doesn't return an array?

I'm happy to get some feedback and move this further.

Cheers,
Robert Scherer

### CODE REVIEW

File: src/AssetConfig.php [hunk 1]
```diff
@@ -525,6 +528,30 @@ public function addTarget($target, array $config)
...
+                $method = $matches[2];
+                $callbackFiles = call_user_func($className . '::' . $method);
+                if (is_array($callbackFiles)) {
+                    // Make sure we insert the files at the correct position, replacing
+                    // the callback string
+                    array_splice($config['files'], $i, 1, $callbackFiles);
+                }
+            }
+        }
+        return $config;
```
Comment (markstory) [inline src/AssetConfig.php]: Other expansions are currently happening in `MiniAsset\Factory::target()`. It current handles remote/local files, will soon handle glob expansion. I see this as similar to glob expansion so perhaps it belongs there?

**Comment by markstory** [comment]
Other than the placement this looks good. There are a few [PHPCS errors](https://travis-ci.org/markstory/mini-asset/jobs/59583393) that I can clean up afterwards if you want.

**Comment by robertscherer** [comment]
Ok - I have moved the functionality to `Factory` and fixed the cs errors.

Preparing the files list outside of the loop has the advantage that callback providers could return a mix of `Local` and `Remote` files. Looking at the glob implementation, the callback providers could return glob patterns and it would work fine.

**Comment by markstory** [comment]
Thanks :+1: I'll update the documentation.

**Pull request merged**